src/views/catalog/wizard/tabs/environment/hook/useEnvironments.ts

```diff
@@ -66,11 +66,10 @@ const useEnvironments = (
   }, [environments, updateVM, vm]);
 
   React.useEffect(() => {
-    const unchanged = environments.every(
-      ({ name, serial }, index) =>
-        initialEnvironments?.[index]?.name === name &&
-        initialEnvironments?.[index]?.serial === serial,
+    const unchanged = environments.every(({ name, serial }) =>
+      initialEnvironments.find((e) => e.name === name && e.serial === serial),
```

Comment:
I'd better use `e?.name` here, same for the `e.serial`. Same for the other code in this PR.